### PR TITLE
wasm-builder: Fix constant re-running of `build.rs` scripts.

### PR DESCRIPTION
**Pull request by bkchr**
Recently we added the wasm binaries to the `rerun-if-changed` list. The problem with that is that
they have a later mtime than the `invoked.timestamp` file and this file's mtime is used to determine
if the `build.rs` script needs to be re-run. The solution to this is that we copy the mtime of this
`invoked.timestamp` file and add it to the wasm binaries. Then cargo/rustc doesn't constantly wants
to rerun the `build.rs` script.

### CODE REVIEW

File: utils/wasm-builder/src/wasm_project.rs [hunk 1]
```diff
@@ -152,9 +152,41 @@ pub(crate) fn create_and_compile(
...
+		build_helper::warning!("Error while adjusting the mtime of the wasm binaries: {}", err)
+	}
+
 	(final_wasm_binary, bloaty)
 }
 
+/// Adjust the mtime of the bloaty and compressed/compact wasm files.
+///
+/// We add the bloaty and the compressed/compact wasm file to the `rerun-if-changed` files.
+/// Cargo/Rustc determines based on the timestamp of the `invoked.timestamp` file that can be found
```
Comment (koute) [inline utils/wasm-builder/src/wasm_project.rs:165]: It's `cargo` that does it; `rustc` doesn't care. (:

**Review by koute — APPROVED** [review]
> Recently we added the wasm binaries to the `rerun-if-changed` list.

What was the motivation for this in the first place?

**Comment by bkchr** [comment]
> > Recently we added the wasm binaries to the `rerun-if-changed` list.
> 
> What was the motivation for this in the first place?

If you deleted the target/debug/wbuild directory it lead to compile errors, because the included files were missing.

**Comment by bkchr** [comment]
Even if we generate the sentinel file, it would be too late. I didn't look into the cargo source code, but it looks like the file needs to exist before the build script is executed. I think the idea is normally that you put everything into the `OUT_DIR`, but this would break windows because the file path would get too long (don't ask), because of this we have this extra `wbuild` directory. This copying of the `mtime` isn't a really great solution, but the best we currently have. In the end the entire `wasm-builder` is a hack :P

**Comment by koute** [comment]
> Even if we generate the sentinel file, it would be too late. I didn't look into the cargo source code, but it looks like the file needs to exist before the build script is executed.

From what I can see it doesn't? (Unless I'm misunderstanding you.) I just tested it out and if I do this in `build.rs`:

```rust
std::fs::write("/tmp/sentinel", "").unwrap();
filetime::set_file_mtime("/tmp/sentinel", 
    filetime::FileTime::from(std::time::SystemTime::now() - std::time::Duration::from_secs(60))
).unwrap();
println!("cargo:rerun-if-changed=/tmp/sentinel");
```

...then the build.rs is just executed once, on the first compilation, and then you need to delete the `/tmp/sentinel` to trigger it again.

(The `set_file_mtime` is necessary since otherwise the whole thing is race-y - I'm assuming because it only checks the timestamp with per-second precision? - so without that it only sometimes works, but if you add it then it works every time from what I can see.)

But, again, I don't really care, so do whatever; the whole thing's janky anyway. 🙈

**Comment by bkchr** [comment]
> > Even if we generate the sentinel file, it would be too late. I didn't look into the cargo source code, but it looks like the file needs to exist before the build script is executed.
> 
> From what I can see it doesn't? (Unless I'm misunderstanding you.) I just tested it out and if I do this in `build.rs`:
> 
> ```rust
> std::fs::write("/tmp/sentinel", "").unwrap();
> filetime::set_file_mtime("/tmp/sentinel", 
>     filetime::FileTime::from(std::time::SystemTime::now() - std::time::Duration::from_secs(60))
> ).unwrap();
> println!("cargo:rerun-if-changed=/tmp/sentinel");
> ```
> 
> ...then the build.rs is just executed once, on the first compilation, and then you need to delete the `/tmp/sentinel` to trigger it again.
> 
> (The `set_file_mtime` is necessary since otherwise the whole thing is race-y - I'm assuming because it only checks the timestamp with per-second precision? - so without that it only sometimes works, but if you add it then it works every time from what I can see.)
> 
> But, again, I don't really care, so do whatever; the whole thing's janky anyway. see_no_evil

Okay xD That doesn't really look any better than my hack xD I mean if it would work without setting the mtime manually, it would be fine, but as you need to do it here as well, I don't see any advantage :P

**Review by andresilva — APPROVED** [review]

**Pull request merged**